### ISSUE TITLE
[0.8] ShadyDOM seems to make it hard to work with other data binding frameworks.

### ISSUE THREAD

**Issue by trading-peter** [Polymer/polymer#1405]
My example is about AngularJS, but I suppose this will happen with any data binding enabled framework. I stumbled upon this when I tried to use `ng-repeat` to generate some `<paper-tab>` elements.

Somewhat like so:

```
var app = angular.module('app', []);

// A simple silly controller
app.controller('myCtrl', function($scope) {
    $scope.tabs = [{name: 'Step 1'}, {name: 'Step 2'}];
});
```

Then somewhere I have some html snippet like this:

```
<div ng-controller="myCtrl">
    <paper-tabs>
        <paper-tab ng-repeat="tab in tabs">{{tab.name}}</paper-tab>
    </paper-tabs>
</div>
```

If you render this while ShadyDOM is in use you'll get something like this:
![Tabs-problem](https://cloud.githubusercontent.com/assets/11567985/7122872/81d9e4c4-e21e-11e4-91e8-f02f46595f1e.jpg)
I added markers because I don't have any styling on my code right now.
The blue marked elements are the tabs rendered in the ShadyDOM, while the red ones are unwanted text node.
I think what happens here is that Polymer builds the ShadyDOM, adding the additional layout into the element but angular is also in the process of doing it's data binding magic and inserts the text node in the somewhat "wrong place" (or not wrong place as far as angular is concerned).

Here's what the markup looks like (first paper-tab expanded):
![developer tools - http___localhost_4443_install](https://cloud.githubusercontent.com/assets/11567985/7123069/cda4035c-e21f-11e4-843c-08591ba09b3f.jpg)

If you enable ShadowDOM this will not happen (of course).

**Comment by schnie** [comment]
@pkaske having similar issues with 1.0 and meteor.  Tried to enable shadow dom rather than shady dom, but having trouble getting it to work in firefox without using the `?dom=shadow` query param.  Curious if you had tested using shadow dom with firefox?

**Comment by pkaske** [comment]
@schnie No I didn't test it in Firefox. I expected the problem to be the exact same as firefox doesn't support native shadow dom out of the box.

**Comment by ebidel** [comment]
https://github.com/PolymerLabs/polymer-experiments/blob/master/patch-dom.html might help. Load it in an html import after Polymer. It patches the DOM methods to work with shady dom.

**Comment by pkaske** [comment]
Thank you. I don't work with angular currently but hopefully it helps the meteor folks.

**Comment by emmanuelbuah** [comment]
@ebidel I know patch-dom is experimental and needs more work. We are struggling with Polymer + other rendering engine because of this issue. It would be great if the polymer team could help us with more work on patch-dom so we can enjoy using polymer with existing rendering engines. We need help.

**Comment by ebidel** [comment]
Can you elaborate what you're missing. Is there something not working?

**Comment by emmanuelbuah** [comment]
@ebidel Sure. I will put together and share a repo for demonstrate rendering issues with meteor.

**Comment by emmanuelbuah** [comment]
@ebidel Per last message, here is a link to publish simple polymer site (using meteor rendering) - http://polymershadow.meteor.com. Polymer is configured to use shadow dom on the page.  Rendering of button on page 1 works fine on load or refresh but has issues rerending if you navigate from page 2 back to page 1. Below are steps to reproduce the issue. 

Steps to reproduce rerending issue 
1. Refrest site
2. Navigate to page 2
3. Navigate back to page 1 

This issue is applicable only to firefox and safari. I haven't tested on Edge or IE. Works fine on chrome but happens on chrome when polymer is configured using shady dom. 

FY- The issue persists after apply patch-dom.html to the page.

**Comment by emmanuelbuah** [comment]
@ebidel I have an app that currently stuck in 0.5 due to this issue. I reported a similar issue at https://github.com/Polymer/polymer/issues/2453#issuecomment-141531275. What can I do to help address this issue? Let me know if there anything I can do to help.

**Comment by ebidel** [comment]
You have patch-dom commented out...? https://github.com/emmanuelbuah/meteorpolymer/blob/master/public/bower_components/vimports.html#L9

**Comment by emmanuelbuah** [comment]
@ebidel I uncommented patch-dom and redeployed to http://polymershadow.meteor.com as well as updating the repo. Everything works fine in chrome and opera but not in firefox or safari.

**Comment by emmanuelbuah** [comment]
@ebidel  I just wanted to say thanks for attending to this ticket. I have decide to drop polymer for my service www.huqoo.com because we've been struggling with 1.0 upgrade. This ticket being one of them. You will find more out meteor + polymer community concerns here - https://gitter.im/emmanuelbuah/MeteorPolymer. In any case, I know you guys are busy and can't attend to every issue. I just wanted to share this. Again, thanks.

**Comment by a4xrbj1** [comment]
Have to agree with the sentiments expressed before like from @emmanuelbuah . The problem of Meteor & Polymer isn't the only framework that's causing problems and yet it still doesn't seem to bother the developers. You guys are missing a golden opportunity to rectify the general view about Polymer based on the 0.5 debacle. If you want developers to talk more positive about Polymer 1.0 then support them and address their problems/issues reported. Especially the major ones like this. Otherwise repeat history and don't wonder why developers have stopped development with Polymer and turned to alternatives as they won't wait forever (yes, there are developers out there that just don't do it for fun but have real paid applications to be build).

**Comment by ebidel** [comment]
> rectify the general view about Polymer based on the 0.5 debacle

Can you explain what this is? We were very clear from the beginning that 0.5 was a developer preview. Despite this, people still built production apps using it, including Google.

> then support them and address their problems/issues reported

We have 2,500+ open issues across Polymer, PolymerLabs, and PolymerElements with dozens of issues reported daily across may different repos. Prioritize becomes the bottleneck. We're addressing critical bugs and performance issues that affect the broadest set of users, first.

I offered a workaround [in August](https://github.com/Polymer/polymer/issues/1405#issuecomment-134771127) and have been asking for more info as comments have come in. Unfortunately, we've been given repo links to full apps, code that has my suggested workaround commented out, and an app (http://polymershadow.meteor.com) where the repo steps do not work in FF/Safari. Clicking the page buttons don't do anything for me. This back and forth increases the amount of noise on the issue and makes it very hard to help debug problems.

If folks on this thread could please post **concrete specifics** on things that are not working, that would be very helpful to move this issue forward.

**Comment by emmanuelbuah** [comment]
@ebidel I agree. There is just a lot of back and forth. I do know both sides want to address the issue. Per my previous message. You have been helpful and continue to be. I appreciate that. I will do my best to clarify.

The sample app hosted at http://polymershadow.meteor.com contains your suggest fix/workaround in August. The fix addresses the issue in chrome and opera but not in safari and firefox. 

The expected result would have been the site (http://polymershadow.meteor.com) works on chrome, opera, safari and firefox. I haven't even run the sample on IE so I don't know if it would work or not. 

In summary, dom-patch addressed the issue in chrome and opera but not in safari and firefox, and I don't know why? 

Let me know if there is any specific information you need to help arrive at the cause for this issue.

**Comment by arthurevans** [comment]
@emmanuelbuah I believe it's working on Chrome and Safari not because of `patch-dom`, but because you're enabling native shadow DOM on Chrome and Opera:

https://github.com/emmanuelbuah/meteorpolymer/blob/master/meteorpolymer.html#L6

The `patch-dom` import doesn't do anything if you have native shadow DOM -- it's not needed:

https://github.com/Polymer/polymer/blob/c2b7c31b8e176a642bbf81f422d19091ed80d5e3/src/lib/experimental/patch-dom.html#L62

When I looked at the Meteor issues before, it appeared that they related to the way that the templating system (Blaze?) applied diffs to the DOM. Digging into it any further was beyond my abilities, but perhaps @sorvell can figure out what about the timing is preventing `patch-dom` from working here (because it appears to me that all of the elements are getting patched correctly in Safari, but I'm guessing somehow the way the elements are created and/or added to the DOM is foiling `patch-dom`.

**Comment by emmanuelbuah** [comment]
@arthurevans @ebidel You guys are awesome. Thanks for your continuing effort. @sorvell Your help to figure out what's foiling patch-dom in relation to this issue will be greatly appreciated.
